BEP60.md

```diff
@@ -0,0 +1,58 @@
+
+# Specification
+
+## Pre-requisites
+
+This BEP requires the existence of an up-to-date SPV node of Binance Chain on the smart-contract-enabled blockchain with which the peg is to be maintained. This means that a smart contract capable of verifying state transitions on Binance Chain should be developed and updated with the constant relaying of transactions that affect it's operations, such as any change in Binance Chain's validator set.
+
+## Trustless Atomic Peg Swap
+
+### Overview
+
+This BEP extends the original BEP3 process in order to remove the parts that require trust in the deputy, which are the parts of the protocol where the deputy acts as a trusted oracle between both chains, relaying information between them. In the current state of BEP3 this is happens in two situations, when new pegged tokens are issued and when they are redeemed.
+
+For token redemption this is pretty straightforward, tokens on Binance Chain can be burned by sending them to the null address and then an SPV proof of this transaction can be relayed by anyone to the smart contract, which will then verify that the transaction actually happened on Binance Chain and send the unlocked tokens to the entity that burned the tokens.
+
+The other case, which involves minting new pegged tokens when real tokens are locked on the smart contract, is much harder to decentralize as there are no smart contracts on Binance Chain, preventing transaction verification on that side of the peg. We get around this by maintaining the figure of the deputy while adding an emergency mechanism, an escape hatch that allows anyone to ring the alarm whenever the deputy behaves incorrectly, prove his bad behavior and freeze everything in order to give all the tokens back to their current rightful owners. After this process has been completed, a new custodian can be selected and the peg restarted.
+
+### Emergency mechanism
+
+The emergency mechanism, which can be triggered by anyone in the event of the custodian minting unbacked coins, follows these steps:
+
+1. Custodian sends tokens to some address that has not previously locked tokens on the smart contract, essentially minting pegged tokens out of thin air and destroying the backing of the peg.
+2. A user finds out about this, creates an SPV proof of the fraudulent transaction described in the previous step and sends it to the smart contract.
+3. The smart contract verifies the validity of the proof and checks that the address where the pegged tokens were sent has no previously locked tokens. If both conditions hold, the contract enters an emergency state, which prevents new tokens from being locked, takes the `AppHash` of the last known good block (the last block before the fraudulent transaction took place) and pins it as the final token distribution, taking a snapshot of all the token balances at that point.
```

Comment:
How can we prove the balance of an address at certain height?

Reply:
According to [the Tendermint docs](https://cosmos-sdk.readthedocs.io/en/stable/ibc.html):
> a field in the block header called AppHash, which refers to the Merkle root hash of the application after processing the transactions from the previous block

`AppHash` is the equivalent in Cosmos to the state root in Ethereum, so it should be possible to construct a merkle path connecting any part of the current state, such as the balance of any account, to `AppHash`.
Verification of this path can then be done in an Ethereum smart contract by walking the IAVL+ path and verifying that each node is correct.